docs/configuration.md
# Mercury Configuration

Mercury uses a [toml](https://github.com/toml-lang/toml) configuration file. this documentation will go through each section to explain the configurable items of each section

## Global Settings

Settings are defined in the `[settings]` block. options are:

Key        | Option                    | Default | Values     | Description
---------- | ------------------------- | ------- | ---------- | -------------------------------------------------------------------------------------------------------------------------
[settings] | manage_network_interfaces | "yes"   | "yes"/"no" | allow mercury to add vip's to the network interfaces - required for internal proxy or for haproxy who does not add vip's.
[settings] | enable_proxy              | "yes"   | "yes"/"no" | use internal proxy for loadbalancing - not needed for external proxy programs, or dns only setup.

## Logging

Log settings are defined in the `[logging]` block. options are:

Key       | Option | Default            | Values                    | Description
--------- | ------ | ------------------ | ------------------------- | ------------------------------------------------------
[logging] | level  | "info"             | "(debug/info/warn/error)" | log level with debug being the most informative.
[logging] | output | "/var/log/mercury" | "(stdout/file)"           | location to write the log information to Web Settings.

## Web

Web interface settings are defined in the `[web]` block. options are:

Key                 | Option   | Default   | Values             | Description
------------------- | -------- | --------- | ------------------ | -------------------------------------------------------------------------------------------------------------------------------------
[web]               | binding  | "0.0.0.0" | string             | ip for the web interface to listen on
[web]               | port     | 9001      | int                | port for the web interface to listen on
[web.tls]           | tls      | none      | see TLS Attributes | TLS certificate information required for SSL
[web.auth.password] | username | none      | sha256 password    | username/password for admin rights in the gui (replace username with the username, and password with the sha256 of a password)
[web.auth.ldap]     | host     | none      | "hostname/ip"      | host of your ldap server
[web.auth.ldap]     | port     | 389       | int                | port of your ldap server
[web.auth.ldap]     | method   | "tls"     | "tls/ssl"          | how to connect to your ldap server, ssl or tls
[web.auth.ldap]     | binddn   | none      | string             | the path to your CN (ex. "OU=Users,DC=example,DC=com"), we will apply the filter to this DN to find the user after authentication
[web.auth.ldap]     | filter   | none      | string             | filter to apply in binddn to find user (%s replaces the username used in login) (ex. "(&(objectClass=organizationalPerson)(uid=%s))")
[web.auth.ldap]     | domain   | none      | string             | the domain to prepend to the username during login
[web.auth.ldap.tls] | tls      | none      | see TLS Attributes | set insecureskipverify = true if required

- note that when enabling LDAP, that local authentication no longer works and that an LDAP authenticated account is required.

## Cluster

Cluster settings are defined in the `[cluster]` block. options are:

Key                | Option                    | Default            | Values                                                       | Description
------------------ | ------------------------- | ------------------ | ------------------------------------------------------------ | ----------------------------------------------------------------------------
[cluster.binding]  | name                      | ""                 | string                                                       | Name of the cluster group
[cluster.binding]  | addr                      | ""                 | string                                                       | ip to bind on for cluster communication
[cluster.binding]  | authkey                   | ""                 | string                                                       | key required to connect to this cluster
[cluster.settings] | connection_timeout        | 10                 | int (seconds)                                                | timeout connecting to remote cluster
[cluster.settings] | connection_retry_interval | 10                 | int (seconds)                                                | time in between retries connecting to the cluster
[cluster.settings] | ping_interval             | 11                 | int (seconds)                                                | how often to send a ping to the remote host
[cluster.settings] | ping_timeout              | 10                 | int (seconds)                                                | host long to wait for a ping timeout (generally 1 second less then interval)
[cluster.settings] | port                      | 9000               | int                                                          | port to listen on for cluster communication
[cluster.tls]      | none                      | see TLS Attributes | TLS certificate information required for SSL
[[cluster.nodes]]  |                           |                    | array of loadbalancer nodes to connect to and form a cluster
[[cluster.nodes]]  | name                      | string             | name of a cluster node
[[cluster.nodes]]  | addr                      | string             | address of a cluster node
[[cluster.nodes]]  | authkey                   | string             | key used to connect to this cluster node

## DNS

DNS settings are defined in the `[dns]` block. options are:

Key   | Option           | Default                                                                                                                   | Values      | Description
----- | ---------------- | ------------------------------------------------------------------------------------------------------------------------- | ----------- | -----------------------------------------------
[dns] | binding          | "0.0.0.0"                                                                                                                 | string      | binding ip for dns service
[dns] | port             | 53                                                                                                                        | int         | binding port for dns service
[dns] | allow_forwarding | []                                                                                                                        | ["ip/mask"] | array of cidrs to allow dns forwarding requests
[dns] | allow_requests   | [ "A", "AAAA", "NS", "MX", "SOA", "TXT", "CAA", "ANY", "CNAME", "MB", "MG", "MR", "WKS", "PTR", "HINFO", "MINFO", "SPF" ] | ["types"]   | array of dns requests types we respond to

## TLS Attributes

TLS attributes are appended to any of the TLS keys in the config.

Usable in the settings for: `cluster`, `web`, `listener`, `backend` and `Healthcheck`

- `[cluster.tls]` - for ssl settings on cluster communication
- `[web.tls]` - for ssl settings on the web gui
- `[loadbalancer.pools.poolname.listener.tls]` - for ssl settings on the pool listener
- `[loadbalancer.pools.poolname.backends.backendname.tls]` - for ssl settings connecting to a backend node with ssl
- `[loadbalancer.pools.poolname.backends.backendname.healthcheck.tls]` - for ssl settings for healthcheck connecting to a backend node with ssl

Key          | Option             | Default        | Values          | Description
------------ | ------------------ | -------------- | --------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
[parent.tls] | minversion         | "VersionTLS12" | string          | Minimum TLS version required for this listener
[parent.tls] | maxversion         | ""             | string          | Maximum TLS version required for this listener
[parent.tls] | ciphersuites       | all            | ["cipher"]      | Cipher suites used by the listener (note that TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 is required for HTTP/2 support. see <https://golang.org/pkg/crypto/tls/#pkg-constants> for details
[parent.tls] | curvepreferences   | all            | ["curve"]       | Curve preference used by the listener. see <https://golang.org/pkg/crypto/tls/#pkg-constants> for details.
[parent.tls] | certificatekey     | ""             | "/path/to/file" | file containing your ssl key
[parent.tls] | certificatefile    | ""             | "/path/to/file" | file containing your ssl certificate file
[parent.tls] | insecureskipverify | false          | true/false      | to to true to ignore insecure certificates, usable for self-signed certificates
[parent.tls] | clientauth         | NoClientCert   | string          | server' policy for client authentication, see <https://golang.org/pkg/crypto/tls/#ClientAuthType> for details

### TLS Min/Max version

Supported versions are:

- VersionSSL30
- VersionTLS10
- VersionTLS11
- VersionTLS12

For details: <https://golang.org/pkg/crypto/tls/#pkg-constants>

### TLS Ciphersuites

Supported ciphersuites are:

- TLS1.2
- TLS_RSA_WITH_RC4_128_SHA
- TLS_RSA_WITH_3DES_EDE_CBC_SHA
- TLS_RSA_WITH_AES_128_CBC_SHA
- TLS_RSA_WITH_AES_256_CBC_SHA
- TLS_RSA_WITH_AES_128_CBC_SHA256
- TLS_RSA_WITH_AES_128_GCM_SHA256
- TLS_RSA_WITH_AES_256_GCM_SHA384
- TLS_ECDHE_ECDSA_WITH_RC4_128_SHA
- TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA
- TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA
- TLS_ECDHE_RSA_WITH_RC4_128_SHA
- TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA
- TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA
- TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA
- TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256
- TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256
- TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
- TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
- TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
- TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
- TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305
- TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305
- TLS1.3
- TLS_AES_128_GCM_SHA256
- TLS_AES_256_GCM_SHA384
- TLS_CHACHA20_POLY1305_SHA256
- Fallback
- TLS_FALLBACK_SCSV

For details: <https://golang.org/pkg/crypto/tls/#pkg-constants>

#### TLS Recommended Cyphers and HTTP/2:

- TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 <- has to be first if you want HTTP/2 support!


CHANGELOG.md
# 1.1.9

misc: fix for pipeline (#129)
pipeline: ensure we check version against the last tag id, the old command could show a older one (#128)

# 1.1.8

misc: changes to the pipeline

# 1.1.7

misc: newer golang version

# 1.1.6

fix:  resolver memory leak (#127)
misc: add optional stackimpact api (#126)
auto: updating change log with latest commit

# 1.1.5

fix: correct api healthcheck concurrency which could get stale on concurrent request (#125)
auto: updating change log with latest commit

# 1.1.4

misc: fix race condition in cluster test (#124)

# 1.1.2

fix: improve healthcheck uuid generation (#123)
misc: allow for port configuration through httprequest healthcheck
fix: correct use of custom ports in health check
fix: correct use of custom ports (#121)
misc: allow for override of port in url request

# 1.1.1

fix: improve ipv6 handling and logging

# 1.1.0

feature: add rule based scripting support (gorules) (#118)
fix: update ldap auth to newer library due to dependency failures
misc: feature: add rule based scripting support (gorules) (#118)

# 1.0.3

misc: Replaced godep with the native version for go modules (#117) (ilijamt)

# 1.0.2

fix: better cname handling with latest resolver update (#116)
misc: update deps to get improved resolver, aswell as improvements in other dependencies
fix: websocket test with new parameters

# 1.0.1

misc: update readme for docker containers
misc: group docker files in dir, and fix version reporting of docker containers (#115)
misc: set git details even if we don't update change log

# 1.0.0

misc: After 2 years of running in production, its time to bump Mercury to a stable version number: 1.0.0
      this version is fully backwards compatible with the previous version, and is only a cosmetic version change

# 0.15.8

feat: TLS 1.3 is now enabled by default on all tls listeners
misc: add security policy (#113)
misc: add docker files, examples and update documentation (#111)
fix: changes to fix tests in golang 1.13 - params not in init anymore

# 0.15.7

misc: fix: prevent paging in ci scripts (#110)
misc: fix codacy issue

# 0.15.5

misc: fix: bump cookbook on new release (#106)
misc: improve readme

# 0.15.4

misc: fix: correct copied dots in ci (#104)

# 0.15.1
Important: X-Forwarded-for now gets automaticly added, if you have X-forwarded-for headers in place, you can remove them. if not, you will have multiple entries in your X-Forwarded-for header

Feature:
  * Show golang version used to compile mercury

Bug:
  * Security fixes in the golang net/http service (recompiled with latest golang vesion)

# 0.15.0
Feature:
  * Allow passing of client certificate using ACLs (lucaslorentz)
  * Adding a sourceip at the listener pool, will force all outgoing connections on that pool to use that as a source ip (use this to prevent martians when hosts have multiple network routes)
  * Added experimental support for TLS1.3 (enabled using GODEBUG=tls13=1, this will be enabled by default when go1.13 is released)

Bug:
  * Default netmask for ipv6 is now /128 instead of /32 (affects network interfaces and topology based loadbalancing checks)

Change:
  * Removed manual websocket implementation for httputil's reverse proxy. it now supports websockets in the main branch
  * DNS forwarder library replaced with one that correctly resolves cnames recursively
  * Error logging of the http/reverseproxy server is now logged in the log file instead of stderr


# 0.14.4
Feature:
  * Improved logging on balance selection in debug mode
  * Improved detailed view of backend
  * Updated all vendor libraries linked to this project
  * Updated to go1.12

Bug:
  * Fix memory leak on non-closing channel (firstbyte received) in the TCP netPipe function used for tcp loadbalancing, when no data was received.

# 0.14.3
Bug:
  * Added a garbage collection on reload due to golang bug, also to track memory usage

# 0.14.2
Bug:
  * Fixed nil pointer issue in modify cookie

# 0.14.1
Bug:
  * Fixed the outbound acl from spawning multiple set-cookies (affected replace and addcookie)

# 0.14.0
Feature:
  * Allow checking of specific parts of the LB using --dns-name, --pool-name, --backend-name and --cluster-only to check specific backends/pools/dns or cluster parts

# 0.13.9
Bug:
  * Health check report had wrong wording when reporting 2 nodes online when 1 was expected

# 0.13.8
Feature:
  * Be able to specify the amount of cluster nodes, or backend nodes for monitoring to alert on
  * Optionally enable golang pprof profiler using the PROFILER_ADDR="localhost:6060" environment variable

# 0.13.7
Feature:
  * Support TLS Client certificates in checks

# 0.13.6
Feature:
  * Added option for REQ_QUERY in ACL url rewriting

# 0.13.5
Change:
  * Added additional security headers to web gui
  * CIDR inbound ACL's now also apply to TCP proxy

# 0.13.4
Feature:
  * Added option to rewrite url before passing it on to the backend
  * Added option to allow/deny requests based on url matches

# 0.13.3
Feature:
  * Added option to use SSH authentication as health check for a node

# 0.13.2
Change:
  * Topology based loadbalancing now returns records based on closest match (instead of all matching records)
 
# 0.13.1
Feature:

internal/config/config.go
package config

import (
	"crypto/sha256"
	"fmt"
	"io/ioutil"
	"os"
	"runtime"
	"strings"
	"sync"
	"time"

	yaml "gopkg.in/yaml.v2"

	"github.com/schubergphilis/mercury/internal/web"
	"github.com/schubergphilis/mercury/pkg/balancer"
	"github.com/schubergphilis/mercury/pkg/cluster"
	"github.com/schubergphilis/mercury/pkg/dns"
	"github.com/schubergphilis/mercury/pkg/healthcheck"
	"github.com/schubergphilis/mercury/pkg/logging"
	"github.com/schubergphilis/mercury/pkg/param"
	"github.com/schubergphilis/mercury/pkg/tlsconfig"

	"github.com/BurntSushi/toml"
)

var (
	config     *Config
	configLock sync.RWMutex
	// Version of application
	Version string
	// VersionBuild number
	VersionBuild string
	// VersionSha git commit of build
	VersionSha string
	// StartTime of application
	StartTime time.Time
	// ReloadTime last time a reload was successfull
	ReloadTime time.Time
	// FailedReloadTime last time a reload failed
	FailedReloadTime time.Time
	// FailedReloadError last time a reload failed
	FailedReloadError string
	// LogLevel holds the level
	LogLevel string
	// LogTarget defines where to write the messages
	LogTarget string
)

const (
	// YES string
	YES            = "yes"
	logConfigLocks = false
)

// Config holds your main config
type Config struct {
	Logging      LoggingConfig `toml:"logging" json:"logging"`
	Cluster      Cluster       `toml:"cluster" json:"cluster"`
	DNS          dns.Config    `toml:"dns" json:"dns"`
	Settings     Settings      `toml:"settings" json:"settings"`
	Loadbalancer Loadbalancer  `toml:"loadbalancer" json:"loadbalancer"`
	Web          web.Config    `toml:"web" json:"web"`
}

// Cluster contains the cluster settings
type Cluster struct {
	Binding   ClusterNode         `toml:"binding" json:"binding"`
	Nodes     []ClusterNode       `toml:"nodes" json:"nodes"`
	Settings  cluster.Settings    `toml:"settings" json:"settings"`
	TLSConfig tlsconfig.TLSConfig `toml:"tls" json:"tls"`
}

// ClusterNode contains the connection details of the cluster node
type ClusterNode struct {
	Name    string `toml:"name" json:"name"`
	Addr    string `toml:"addr" json:"addr"`
	AuthKey string `toml:"authkey" json:"authkey"`
}

// Settings contains a list of global application settings
type Settings struct {
	ManageNetworkInterfaces string `toml:"manage_network_interfaces"` // do network interface config (e.g. bind ip's)
	EnableProxy             string `toml:"enable_proxy"`              // start proxies, or let another app handle this
}

// LoggingConfig log config
type LoggingConfig struct {
	Level  string `toml:"level" json:"level"`
	Output string `toml:"output" json:"output"`
}

// ReloadConfig reloads the config (typically after a hup)
func ReloadConfig() {
	log := logging.For("config/reload")
	log.Info("Reload config file")
	err := LoadConfig(*param.Get().ConfigFile)
	if err != nil {
		FailedReloadTime = time.Now()
		FailedReloadError = err.Error()
		log.Warnf("Error reloading config:%s", err)
	} else {
		ReloadTime = time.Now()
		log.Infof("Reloaded")
	}
}

// LoadConfig a config file
func LoadConfig(file string) error {
	log := logging.For("config/load")
	log.Info("Loading config")
	data, err := ioutil.ReadFile(file)
	if err != nil {
		return err
	}
	temp := new(Config)
	f := strings.Split(file, ".")
	switch f[len(f)-1] {
	case "toml":
		log.Debug("Decode toml config")
		_, err = toml.Decode(string(data), temp)
		if err != nil {
			return err
		}
	case "yaml":
		log.Debug("Decode yaml config")
		err = yaml.Unmarshal([]byte(data), temp)
		if err != nil {
			return err
		}
	}

	log.Debug("Check config")
	if err = temp.ParseConfig(); err != nil {
		return err
	}

	log.Debug("Activating new config")
	configLock.Lock()
	config = temp

	log.Info("Config loaded succesfully")
	configLock.Unlock()

	return nil
}

// ParseConfig parses the config and returns an error if failed
func (c *Config) ParseConfig() error {
	log := logging.For("config/parse")
	// Check SSL Certificates
	err := c.ValidateCertificates()
	if err != nil {
		return err
	}

	// Loadbalance defaults
	if c.Loadbalancer.Settings.DefaultLoadBalanceMethod == "" {
		c.Loadbalancer.Settings.DefaultLoadBalanceMethod = "roundrobin"
	}

	ipportmap := make(map[string]string)

	// Ensure a default in all backends
	for poolName, pool := range c.Loadbalancer.Pools {
		if pool.ErrorPage.File != "" {
			if _, err := os.Stat(pool.ErrorPage.File); err != nil {
				return fmt.Errorf("Cannot access error page for pool:%s file:%s error:%s", poolName, pool.ErrorPage.File, err)
			}
		}

		if pool.MaintenancePage.File != "" {
			if _, err := os.Stat(pool.MaintenancePage.File); err != nil {
				return fmt.Errorf("Cannot access maintenance page for pool:%s file:%s error:%s", poolName, pool.MaintenancePage.File, err)
			}
		}

		p := c.Loadbalancer.Pools[poolName]
		if p.ErrorPage.TriggerThreshold == 0 {
			p.ErrorPage.TriggerThreshold = 500
		}

		if p.Listener.Mode == "" {
			p.Listener.Mode = "tcp"
		}

		var protocol string
		switch p.Listener.Mode {
		case "tcp", "http", "https":
			protocol = "tcp"
		case "udp":
			protocol = "udp"
		}

		if p.Listener.IP != "" {
			portmap := fmt.Sprintf("%s_%d_%s", p.Listener.IP, p.Listener.Port, protocol)
			if conflict, ok := ipportmap[portmap]; ok {
				return fmt.Errorf("Duplicate ip/port configuration: both pools are using the same ip/port: %s and %s use %s:%d (%s/%s)", poolName, conflict, p.Listener.IP, p.Listener.Port, protocol, p.Listener.Mode)
			}
			ipportmap[portmap] = poolName
		}

		if p.Listener.OCSPStapling == "" {
			p.Listener.OCSPStapling = YES
		}

		if p.Listener.MaxConnections == 0 {
			p.Listener.MaxConnections = 2048
		}

		if p.Listener.HTTPProto == 0 {
			p.Listener.HTTPProto = 2
		}

		// Default writetimeout for listener is 0 = unlimited time
		// Default readtimeout for listener is 10 seconds
		if p.Listener.ReadTimeout == 0 {
			p.Listener.ReadTimeout = 10
		}

		c.Loadbalancer.Pools[poolName] = p

		for hid, check := range c.Loadbalancer.Pools[poolName].HealthChecks {
			p := c.Loadbalancer.Pools[poolName]
			p.HealthChecks[hid] = SetHealthCheckDefault(check)
			if p.Listener.SourceIP != "" {
				p.HealthChecks[hid].SourceIP = p.Listener.SourceIP
			}
			c.Loadbalancer.Pools[poolName] = p
		}

		for backendName, backend := range c.Loadbalancer.Pools[poolName].Backends {
			h := backend

			if backend.UUID == "" {
				// generate hash uniq to cluster - pool - backend
				hash := sha256.New()
				hash.Write([]byte(fmt.Sprintf("%s-%s-%s", c.Cluster.Binding.Addr, poolName, backendName)))
				h.UUID = fmt.Sprintf("%x", hash.Sum(nil))
			}

			if backend.ConnectMode == "" {
				h.ConnectMode = c.Loadbalancer.Pools[poolName].Listener.Mode
			}

			if backend.DNSEntry.IP == "" && c.Loadbalancer.Pools[poolName].Listener.IP == "" {
				return fmt.Errorf("No IP defined in either the pool's listener IP or the DNSentry IP for backend:%s", backendName)
			}
			// If not DNS Entry IP is set, set the ip to the listener
			if backend.DNSEntry.IP == "" {
				h.DNSEntry.IP = c.Loadbalancer.Pools[poolName].Listener.IP
			}

			if backend.ErrorPage.File != "" {
				if _, err := os.Stat(backend.ErrorPage.File); err != nil {
					return fmt.Errorf("Cannot access error page for pool:%s backend:%s file:%s error:%s", poolName, backendName, backend.ErrorPage.File, err)
				}
			}

			if backend.ErrorPage.TriggerThreshold == 0 {
				h.ErrorPage.TriggerThreshold = 500
			}

			if backend.MaintenancePage.File != "" {
				if _, err := os.Stat(backend.MaintenancePage.File); err != nil {
					return fmt.Errorf("Cannot access maintenance page for pool:%s backend:%s file:%s error:%s", poolName, backendName, backend.MaintenancePage.File, err)
				}
			}

			for hid, check := range c.Loadbalancer.Pools[poolName].Backends[backendName].HealthChecks {
				h.HealthChecks[hid] = SetHealthCheckDefault(check)
				if backend.BalanceMode.ActivePassive == YES {
					h.HealthChecks[hid].ActivePassiveID = backend.UUID
				} else {
					h.BalanceMode.ActivePassive = "no"
					h.HealthChecks[hid].ActivePassiveID = ""
				}
				if p.Listener.SourceIP != "" {
					h.HealthChecks[hid].SourceIP = p.Listener.SourceIP
				}
			}

			// Always have atleast 1 check: tcpconnect
			if len(c.Loadbalancer.Pools[poolName].Backends[backendName].HealthChecks) == 0 {
				tcpconnect := SetHealthCheckDefault(healthcheck.HealthCheck{})
				if p.Listener.SourceIP != "" {
					tcpconnect.SourceIP = p.Listener.SourceIP
				}
				h.HealthChecks = append(h.HealthChecks, tcpconnect)
			}

			if backend.HealthCheckMode == "" {
				h.HealthCheckMode = "all"
			}

			// Backwards compatibility: if ClusterNodes is set, put this in the new ServingClusterNdoes
			if backend.BalanceMode.ClusterNodes != 0 {
				h.BalanceMode.ServingClusterNodes = backend.BalanceMode.ClusterNodes
			}

			if backend.BalanceMode.ServingClusterNodes == 0 {
				h.BalanceMode.ServingClusterNodes = len(c.Cluster.Nodes)
			}

			if backend.BalanceMode.ServingBackendNodes == 0 {
				h.BalanceMode.ServingBackendNodes = len(backend.Nodes)
			}

			if backend.BalanceMode.LocalTopology != "" {
				if val, ok := c.Loadbalancer.Networks[backend.BalanceMode.LocalTopology]; ok {
					for _, network := range val.CIDRs {
						h.BalanceMode.LocalNetwork = append(h.BalanceMode.LocalNetwork, network)
					}
				} else {
					return fmt.Errorf("Could not find topology name:%s in the defined loadbalancer networks in the config for backend:%s", backend.BalanceMode.LocalTopology, backendName)
				}
			}

			// Default node settings
			for nodeID, node := range c.Loadbalancer.Pools[poolName].Backends[backendName].Nodes {
				if node.UUID == "" {
					// generate hash uniq to pool - backend - node + port (cluster pool removed for stickyness across clusters)
					hash := sha256.New()
					hash.Write([]byte(fmt.Sprintf("%s-%s-%s-%s-%d-%d", poolName, backendName, node.SafeName(), node.Hostname, node.Preference, node.Weight)))

					n := node
					n.UUID = fmt.Sprintf("%x", hash.Sum(nil))
					n.ClusterName = c.Cluster.Binding.Name
					if n.MaxConnections == 0 {
						n.MaxConnections = pool.Listener.MaxConnections
					}
					h.Nodes[nodeID] = n
					log.Infof("Node:%s UUID:%s", h.Nodes[nodeID].Name(), h.Nodes[nodeID].UUID)
				}
				if node.LocalTopology != "" {
					if val, ok := c.Loadbalancer.Networks[node.LocalTopology]; ok {
						for _, network := range val.CIDRs {
							c.Loadbalancer.Pools[poolName].Backends[backendName].Nodes[nodeID].LocalNetwork = append(c.Loadbalancer.Pools[poolName].Backends[backendName].Nodes[nodeID].LocalNetwork, network)
						}
					} else {
						return fmt.Errorf("Could not find topology name:%s in the defined loadbalancer networks in the config for backend:%s node:%s", backend.BalanceMode.LocalTopology, backendName, node.Name())
					}
				}
			}

			// Save Backend changes
			c.Loadbalancer.Pools[poolName].Backends[backendName] = h

			// Copy node Status if exists
			if Get() != nil {
				log.Debug("Config is not empty, copying node status if it still exists")
				if _, ok := Get().Loadbalancer.Pools[poolName]; ok {
					log.WithField("poolname", poolName).Debug("Existing pool")
					if _, ok := Get().Loadbalancer.Pools[poolName].Backends[backendName]; ok {
						log.WithField("poolname", poolName).WithField("backendname", backendName).Debug("Existing backend")
						for _, oldnode := range Get().Loadbalancer.Pools[poolName].Backends[backendName].Nodes {
							for nid, newnode := range c.Loadbalancer.Pools[poolName].Backends[backendName].Nodes {
								if oldnode.UUID == newnode.UUID {
									c.Loadbalancer.Pools[poolName].Backends[backendName].Nodes[nid].Status = oldnode.Status
									c.Loadbalancer.Pools[poolName].Backends[backendName].Nodes[nid].Errors = oldnode.Errors
									log.Debugf("Old node:%s uuid:%s copied to New node:%s uuid:%s", oldnode.Name(), oldnode.UUID, newnode.Name(), newnode.UUID)
								}
							}
						}
					}
				}
			}

		}
	}

	SetDefaultSettingsConfig(&c.Settings)
	SetDefaultClusterConfig(&c.Cluster.Settings)
	SetDefaultDNSConfig(&c.DNS)
	SetDefaultWebConfig(&c.Web)
	return nil
}

// SetHealthCheckDefault sets the default config for generic settings
func SetHealthCheckDefault(check healthcheck.HealthCheck) healthcheck.HealthCheck {
	if check.Interval < 1 {
		check.Interval = 11
	}

	if check.Timeout < 1 {
		check.Timeout = 10
	}

	if check.PINGpackets == 0 {
		check.PINGpackets = 4
	}

	if check.PINGtimeout == 0 {
		check.PINGtimeout = 1
	}

	if check.Type == "" {
		check.Type = "tcpconnect"
	}

	if check.HTTPFollowRedirect == "" {
		check.HTTPFollowRedirect = "yes"
	}

	switch check.OnlineState.Status {
	case healthcheck.Automatic:
		check.OnlineState.Status = healthcheck.Online
	}

	switch check.OfflineState.Status {
	case healthcheck.Automatic:
		check.OfflineState.Status = healthcheck.Offline
	}

	return check
}

// SetDefaultSettingsConfig sets the default config for generic settings
func SetDefaultSettingsConfig(s *Settings) {
	if s.ManageNetworkInterfaces == "" {
		s.ManageNetworkInterfaces = YES
	}

	if s.EnableProxy == "" {
		s.EnableProxy = YES
	}
}

// SetDefaultClusterConfig sets the default config for ClusterService
func SetDefaultClusterConfig(d *cluster.Settings) {
	if d.ConnectInterval < 1*time.Second {
		d.ConnectInterval = 10
	}

	if d.ConnectTimeout < 1*time.Second {
		d.ConnectTimeout = 10
	}

	if d.PingInterval < 1*time.Second {
		d.PingInterval = 5
	}

	if d.ReadTimeout < 1*time.Second {
		d.ReadTimeout = 11
	}

	if d.JoinDelay < 1 {
		d.JoinDelay = 500 * time.Microsecond
	}
}

// SetDefaultDNSConfig sets the default config for DNSService
func SetDefaultDNSConfig(d *dns.Config) {
	if d.Binding == "" {
		d.Binding = "localhost"
	}

	if d.Port < 1 {
		d.Port = 53
	}

	if len(d.AllowedRequests) == 0 {
		// Allow the most common DNS request types
		d.AllowedRequests = []string{"A", "AAAA", "NS", "MX", "SOA", "TXT", "CAA", "ANY", "CNAME", "MB", "MG", "MR", "WKS", "PTR", "HINFO", "MINFO", "SPF"}
	}

	for domainName, localDomain := range d.Domains {
		for rid, record := range localDomain.Records {

			// Add UUID to static records, dynamic ones are auto-generated
			if record.Statistics == nil {
				hash := sha256.New()
				hash.Write([]byte(fmt.Sprintf("%s-%s-%x-%s", domainName, record.Name, record.Type, record.Target)))
				uuid := fmt.Sprintf("%x", hash.Sum(nil))
				d.Domains[domainName].Records[rid].UUID = uuid
				d.Domains[domainName].Records[rid].Statistics = balancer.NewStatistics(uuid, 0)
				if d.Domains[domainName].Records[rid].LocalNetwork != "" {
					d.Domains[domainName].Records[rid].Statistics.Topology = []string{d.Domains[domainName].Records[rid].LocalNetwork}
				}
			}
		}
	}
}

// SetDefaultWebConfig sets the default config for Webservice
func SetDefaultWebConfig(w *web.Config) {
	// Create empty user hash if not defined
	if w.Auth.Password == nil {
		w.Auth.Password = &web.AuthPassword{
			Users: make(map[string]string),
		}
	}

	if w.Binding == "" {
		w.Binding = "localhost"
	}

	if w.Port == 0 {
		w.Port = 9001
	}

	// Set default LDAP settings
	if w.Auth.LDAP != nil {
		if w.Auth.LDAP.Method == "" {
			w.Auth.LDAP.Method = "TLS"
		}
		if w.Auth.LDAP.Port == 0 {
			w.Auth.LDAP.Port = 389
		}
	}
}

// ValidateCertificates checks if all provided SSL certificates are correct
func (c *Config) ValidateCertificates() error {
	// Test Pool/Backend Certificate
	for poolName, pool := range c.Loadbalancer.Pools {
		if strings.EqualFold(pool.Listener.Mode, "https") {

			// Check if we have a main certificate FILE
			certcount := 0
			if pool.Listener.TLSConfig.CertificateProvided() {
				if err := pool.Listener.TLSConfig.Valid(); err != nil {
					return fmt.Errorf("Certificate issue for pool:%s %s", poolName, err.Error())
				}

				certcount++
			}

			// Check if we have certificates on a backend
			for backendName, backend := range pool.Backends {
				if backend.TLSConfig.CertificateProvided() {
					if err := pool.Listener.TLSConfig.Valid(); err != nil {
						return fmt.Errorf("Certificate issue for pool:%s backend:%s %s", poolName, backendName, err.Error())
					}

					for _, check := range backend.HealthChecks {
						if check.TLSConfig.CertificateProvided() {
							if err := check.TLSConfig.Valid(); err != nil {
								return fmt.Errorf("Certificate issue for pool:%s backend:%s check:%s_%s_%d %s", poolName, backendName, check.Type, check.IP, check.Port, err.Error())
							}
						}
					}

					certcount++
				}
			}

			if certcount == 0 {
				return fmt.Errorf("No certificate file specified for HTTPS mode on pool %s", poolName)
			}
		}
	}

	// Test Web Service Certificate
	if c.Web.TLSConfig.CertificateProvided() {
		if err := c.Web.TLSConfig.Valid(); err != nil {
			return fmt.Errorf("Could not load TLS configuration for Mercury Web Service: %s", err)
		}
	}

	// Test Cluster Service Certificate
	if c.Cluster.TLSConfig.CertificateProvided() {
		if err := c.Cluster.TLSConfig.Valid(); err != nil {
			return fmt.Errorf("Could not load TLS configuration for Mercury Cluster Service: %s", err)
		}
	}
	return nil
}

// Get returns the pointer to the latest config loaded
func Get() *Config {
	log := logging.For("config/get")
	_, file, no, ok := runtime.Caller(1)
	if ok && logConfigLocks {
		log.Debugf("getconfig for %s#%d", file, no)
	}

	configLock.RLock()
	defer configLock.RUnlock()
	return config
}

// GetNoLock is Pointer to config without locking
func GetNoLock() *Config {
	return config
}

// Lock the config for Writes
func Lock() {
	log := logging.For("config/lock")
	_, file, no, ok := runtime.Caller(1)
	if ok && logConfigLocks {
		log.Debugf("lockconfig for %s#%d", file, no)
	}
	configLock.Lock()
}

// Unlock the config for Writes
func Unlock() {
	log := logging.For("config/unlock")
	_, file, no, ok := runtime.Caller(1)
	if ok && logConfigLocks {
		log.Debugf("unlockconfig for %s#%d", file, no)
	}
	configLock.Unlock()
}

// RLock the config for reads
func RLock() {
	log := logging.For("config/rlock")
	_, file, no, ok := runtime.Caller(1)
	if ok && logConfigLocks {
		log.Debugf("rlockconfig for %s#%d", file, no)
	}
	configLock.RLock()
}

// RUnlock the config for reads
func RUnlock() {
	log := logging.For("config/runlock")
	_, file, no, ok := runtime.Caller(1)
	if ok && logConfigLocks {
		log.Debugf("runlockconfig for %s#%d", file, no)
	}
	configLock.RUnlock()
}
